## Question1-Answer

In [22]:
A = [1, 2, 3, 4, 5, 6] # creating list A
B = [13, 21, 34] # creating list B

In [18]:
A_B = [1, 2, 3, 4, 5, 6, 13, 21, 34] #Creating list A_B

In [19]:
A.extend(B) # Extending list A with list B

In [20]:
A #show output of A after extending

[1, 2, 3, 4, 5, 6, 13, 21, 34]

In [21]:
assert A_B == A #  Checking if it's true that A.extend(B) is the same list as A_B

**Therefore A_B = [1, 2, 3, 4, 5, 6, 13, 21, 34] can be gotten by the option of A.extend(B)**

## Question2-Answer

In [24]:
import numpy as np

In [25]:
f = np.identity(3)

In [26]:
print(f)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


**Therefore np.identity(3) can be used to create an identity matrix in python**

## Question3-Answer

In [29]:
import pandas as pd

In [30]:
utility_info_df = pd.read_csv('https://bit.ly/HDSC-StageOneDataset')

In [31]:
utility_info_df.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

In [32]:
desc_data = utility_info_df.groupby(['fuel_type_code_pudl'])['fuel_cost_per_unit_burned'].describe()

In [33]:
desc_data

count         mean           std     min      25%  \
fuel_type_code_pudl                                                        
coal                  8547.0    67.421830    943.771433    0.00  23.9585   
gas                  11486.0    13.659397    163.286631    0.00   3.2300   
nuclear                818.0  4955.157002  14291.929749    0.00   1.6015   
oil                   8064.0   168.877086   2406.101479    0.00  37.6100   
other                  167.0    18.253856     29.016882 -119.04   2.6325   
waste                  441.0    19.518122     29.988017 -276.08  10.3350   

                         50%       75%         max  
fuel_type_code_pudl                                 
coal                 33.3180  44.66800   44033.320  
gas                   4.6500   6.96000   10490.000  
nuclear              32.5200  50.85525   69785.215  
oil                  68.9455  98.68925  139358.000  
other                 9.0640  20.05500     134.547  
waste                21.9330  32.65000     132.291

In [34]:
desc_data['mean'].min()

13.659396569737135

**Therefore the fuel type code that has lowest mean value cost per unit burned is gas**

## Question4-Answer

In [35]:
desc_data2 = utility_info_df['fuel_mmbtu_per_unit'].describe()

In [36]:
desc_data2

count    29523.000000
mean         8.492111
std         10.600220
min          0.000001
25%          1.024000
50%          5.762694
75%         17.006000
max        341.260000
Name: fuel_mmbtu_per_unit, dtype: float64

In [37]:
desc_data2['std'].round(2)

10.6

In [38]:
desc_data2['75%'].round(2)

17.01

**Therefore the standard deviation and 75% percentile of fuel_mmbtu_per_unit in two decimal place is 10.60 and 17.01**

## Question5-Answer

In [39]:
utility_info_df['fuel_qty_burned'].skew().round(2)

15.85

In [40]:
utility_info_df['fuel_qty_burned'].kurtosis().round(2)

651.37

**Therefore the skewness and kurtosis for the 'fuel_qty_burned' is 15.85 and 651.37 respectively**

In [41]:
utility_info_df.isna().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [42]:
Total = utility_info_df['fuel_unit'].isna().sum()
Total

180

In [43]:
total_rc = len(utility_info_df['fuel_unit'])
total_rc

29523

In [46]:
Percentage = ((Total / total_rc)).round(3)
Percentage

0.006

**Therefore: Feature:fuel_unit, Total:180, Percentage:0.006**

## Question7-Answer

In [50]:
utility_info_df['fuel_unit'].unique()

array(['ton', 'mcf', 'kgU', 'bbl', 'gramsU', nan, 'mwdth', 'mmbtu',
       'mwhth', 'gal'], dtype=object)

**The column with missing value is a Categorical data and can use mode imputation techniques for missing values**

## Question8-Answer

In [56]:
Corr = utility_info_df.corr()['fuel_cost_per_unit_burned']
Corr

utility_id_ferc1               -0.037863
report_year                     0.013599
fuel_qty_burned                -0.018535
fuel_mmbtu_per_unit            -0.010034
fuel_cost_per_unit_burned       1.000000
fuel_cost_per_unit_delivered    0.011007
fuel_cost_per_mmbtu            -0.000437
Name: fuel_cost_per_unit_burned, dtype: float64

In [59]:
Corr.sort_values()[1:3]

fuel_qty_burned       -0.018535
fuel_mmbtu_per_unit   -0.010034
Name: fuel_cost_per_unit_burned, dtype: float64

**Therefore the features with the second and third lowest correlation to the fuel cost per unit burned are: fuel_qty_burned and fuel_mmbtu_per_unit**

## Question9-Answer

In [60]:
utility_info_df.groupby(['report_year', 'fuel_type_code_pudl'])['fuel_cost_per_unit_burned'].sum().head(30)

report_year  fuel_type_code_pudl
1994         coal                    14984.572
             gas                     10792.496
             nuclear                227983.354
             oil                      9362.194
             other                     116.737
             waste                      52.762
1995         coal                    14571.785
             gas                      9087.905
             nuclear                384152.639
             oil                     17858.281
             other                     183.577
             waste                    -192.699
1996         coal                    12694.803
             gas                      1467.390
             nuclear                 66657.887
             oil                      5357.028
             other                     160.075
             waste                      88.092
1997         coal                    43742.178
             gas                      1518.471
             nuclear       

In [62]:
Percentage_change = ((11902.597 - 14984.572)/14984.572) * 100
round(Percentage_change)

-21

**Therefore the percentage change in fuel cost for the fuel type coal per unit burned in 1998 compared to 1994 is negative and approximately 21 percent**

## Question10-Answer

In [63]:
grouped_data = utility_info_df.groupby(['report_year'])['fuel_cost_per_unit_delivered'].mean()
grouped_data

report_year
1994       63.636060
1995       32.735269
1996     9196.705948
1997    11140.197239
1998      287.154420
1999       25.551627
2000      985.362877
2001       60.050396
2002       47.594361
2003       55.663493
2004      139.524275
2005       41.438184
2006       38.657484
2007       43.325023
2008       58.588197
2009      652.694163
2010       91.862105
2011       59.774667
2012       60.994502
2013      172.307591
2014      192.737183
2015      326.535511
2016      103.901761
2017       46.196861
2018      499.269966
Name: fuel_cost_per_unit_delivered, dtype: float64

In [64]:
grouped_data.max()

11140.19723948813

**Therefore, 1997 is the year with the highest average fuel cost per unit delivered**